architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,442 @@
+
+Dependency: Stage 1
+
+Exit criteria: all functional tests passing using new routing behavior in UCP. All built-in resource types can be queried with the CLI.
+
+#### Stage 3: CLI enhancements
+
+- Remove hardcoded resource type validation from the CLI and replace with queries of the resource provider/type APIs.
+- Add support for `rad resource create`. A new CLI command that can create resources of any type. This is needed to properly exercise the user-defined types feature until Bicep support comes online.
+
+Dependency: Stage 1
+
+Exit criteria: existing tests passing. Add functional test for `rad resource create` using an existing built-in resource.
+
+#### Stage 4: Dynamic Resource Provider
+
+- Implement dynamic resource provider with support for CRUDL operations.
+  - No support for validation (yet).
+  - The dynamic resource provider requires the use of Recipes.
+- Implement integration tests and functional tests for Dynamic RP's basic resource lifecycle using `rad resource create`.
+
+Dependency: Stage 2
+
+Exit criteria: new functional tests exercise the dynamic resource provider.
+
+#### Stage 5: Bicep Support
+
+- Implement a TypeSpec generator that can generate a resource provider manifest and Bicep extension manifest.
+- (If necessary): Update deployment engine logic to generate correct resource ids/URLs for user-defined types.
+- Implement functional tests that exercise user-defined types with Bicep.
+
+Dependency: Stage 4 needed for functional testing, no dependency to begin work.
+
+Exit criteria: new functional tests that exercise the dynamic resource provider with Bicep. 
+
+### Phase 2: Completing the UDT feature
+
+Note: at this point the UDT feature can do useful work end-to-end. The stages in this section make UDT feature-complete.
+
+#### Stage 6: Schema Validation
+
+- Accept an OpenAPI v3 schema as part of a resource type.
+- Implement OpenAPI validation in the dynamic resource provider.
+
+Dependency: Stage 4
+
+Exit criteria: new negative tests for the dynamic resource provider.
+
+#### Stage 7: Safe Unregistration/Modification
+
+- Polish support for unregistering a resource provider.
+  - An in-progress `DELETE` operation on a resource type should block operations on resources of that type other than `DELETE`.
+  - An in-progress `PUT` operation on a resource type should block all operations on resources of that type.
+  - When a resource type is being `DELETE`ed it should also delete all resources of that type.
+
+Dependency: Stage 4
+
+Exit criteria: new negative tests for the dynamic resource provider.
+
+#### Stage 8: Implement connections
+
+- The container RP should be able to define connections to a user-defined type. 
+  - Connections includes app-graph.
+  - Connections includes injecting configuration values and secrets into the container.
+
+Dependency: Stage 4
+
+Exit criteria: new functional tests the connections scenario.
+
```

Comment:
should there be stage for publishing and contributing the types to upstream as highlighted in the user experience which covers documentation, community repo etc

Reply:
This is **our** implementation plan, we are the upstream 😆 

I'll add documentation to this.

For the "contrib" repo, I can't make an implementation plan until we know the details.

Reply:
Meant more of what should Radius do/provide to enable contribution of types to upstream. Makes sense to flush it after we know more.